File: Learning/Python/Python数据结构与算法/算法/排序/4希尔排序/Shellsort.py
#希尔排序法
#将待排序数组分割 再插入排序

def ShellSort(k):
    gap = n = len(k)

    while(gap!=1):
        gap = gap//3 + 1
        for i in range(gap,n):
            if k[i] < k[i-gap]:
                temp = k[i]
                j = i - gap
                while( k[j]>temp and j>=0 ):
                    k[j+gap] = k[j]
                    j -= gap
                k[j+gap] = temp

                
k = [55, 2, 8, 1, 34, 21, 5, 3, 89, 13]
ShellSort(k)

print("从小到大排列为:")
for i in range(len(k)):
    print(k[i], end=' ')
            

File: Learning/数据结构与算法/算法/排序/4希尔排序/ShellSort.c
#include <stdio.h>

void ShellSort(int k[], int n)
{
    int i, j, temp;
    int gap=n;
    int count1=0, count2=0;

    do
    {
        gap = gap/3 + 1;
        for(i=gap; i<n; i++)
        {
            if( k[i] < k[i-gap] )
            {
                count1++;
                temp = k[i];
                for(j=i-gap; k[j]>temp && j>=0; j-=gap) //!j>=0
                {
                    count2++;
                    k[j+gap] = k[j];
                }
                k[j+gap] = temp;
            }
        }
    }while(gap>1);
    printf("总共进行了%d次比较，%d次移动\n",count1, count2);
}

int main()
{
    int i, a[]={5,2,6,0,3,9,1,7,4,8};
    int len=sizeof(a)/sizeof(int);
    ShellSort(a, len);
    printf("排序后的结果是:\n");
    for(i=0; i<len; i++)
    {
        printf("%d ", a[i]);
    }
    printf("\n\n");
}
